src/analysis/inner_product_space/l2_space.lean

```diff
@@ -242,69 +249,134 @@ begin
+  (hV : is_hilbert_sum 𝕜 E V) (W₀ : Π₀ (i : ι), G i) :
+  hV.linear_isometry_equiv.symm (W₀.sum (lp.single 2)) = (W₀.sum (λ i, V i)) :=
+by simp [is_hilbert_sum.linear_isometry_equiv,
   orthogonal_family.linear_isometry_apply_dfinsupp_sum_single]
 
-/-- In the canonical isometric isomorphism `E ≃ₗᵢ[𝕜] lp G 2` induced by an `ι`-indexed orthogonal
-family `G`, a finitely-supported vector in `lp G 2` is the image of the associated finite sum of
+/-- In the canonical isometric isomorphism between a Hilbert sum `E` of `G : ι → Type*` and
+`lp G 2`, a finitely-supported vector in `lp G 2` is the image of the associated finite sum of
 elements of `E`. -/
-@[simp] protected lemma linear_isometry_equiv_apply_dfinsupp_sum_single
-  [Π i, complete_space (G i)]
-  (hV' : ⊤ ≤ (⨆ i, (V i).to_linear_map.range).topological_closure) (W₀ : Π₀ (i : ι), G i) :
-  (hV.linear_isometry_equiv hV' (W₀.sum (λ i, V i)) : Π i, G i) = W₀ :=
+@[simp] protected lemma is_hilbert_sum.linear_isometry_equiv_apply_dfinsupp_sum_single
+  (hV : is_hilbert_sum 𝕜 E V) (W₀ : Π₀ (i : ι), G i) :
+  (hV.linear_isometry_equiv (W₀.sum (λ i, V i)) : Π i, G i) = W₀ :=
 begin
-  rw ← hV.linear_isometry_equiv_symm_apply_dfinsupp_sum_single hV',
+  rw ← hV.linear_isometry_equiv_symm_apply_dfinsupp_sum_single,
   rw linear_isometry_equiv.apply_symm_apply,
   ext i,
   simp [dfinsupp.sum, lp.single_apply] {contextual := tt},
 end
 
-end orthogonal_family
+/-- Given a total orthonormal family `v : ι → E`, `E` is a Hibert sum of `λ i : ι, 𝕜` relative to
+the family of linear isometries `λ i, λ k, k • v i`. -/
+lemma orthonormal.is_hilbert_sum {v : ι → E} (hv : orthonormal 𝕜 v)
+  (hsp : ⊤ ≤ (span 𝕜 (set.range v)).topological_closure) :
+  @is_hilbert_sum _ 𝕜 _ _ _ _ (λ i : ι, 𝕜) _
+    (λ i, linear_isometry.to_span_singleton 𝕜 E (hv.1 i)) :=
+hv.orthogonal_family.is_hilbert_sum
+begin
+  convert hsp,
+  simp [← linear_map.span_singleton_eq_range, ← submodule.span_Union],
+end
+
+lemma direct_sum.is_internal.is_hilbert_sum {F : ι → submodule 𝕜 E}
+  [Π i, complete_space (F i)] (hF : direct_sum.is_internal F)
+  (hF' : @orthogonal_family 𝕜 _ _ _ _ (λ i, F i) _ (λ i, (F i).subtypeₗᵢ)) :
+  @is_hilbert_sum _ 𝕜 _ _ _ _ (λ i, F i) _ (λ i, (F i).subtypeₗᵢ) :=
+hF'.is_hilbert_sum_internal _ (by rw hF.submodule_supr_eq_top; exact subset_closure)
+
+lemma submodule.is_hilbert_sum_orthogonal (K : submodule 𝕜 E) [hK : complete_space K] :
+  @is_hilbert_sum _ 𝕜 _ E _ _ (λ b, ((cond b K Kᗮ : submodule 𝕜 E) : Type*)) _
+  (λ b, (cond b K Kᗮ).subtypeₗᵢ) :=
+begin
+  haveI : Π b, complete_space ((cond b K Kᗮ : submodule 𝕜 E) : Type*),
+  { intro b,
+    cases b;
+    exact orthogonal.complete_space K <|> assumption },
+  refine direct_sum.is_internal.is_hilbert_sum _ K.orthogonal_family_self,
+  convert (direct_sum.is_internal_submodule_iff_is_compl _ (by contradiction) bool.univ_eq).mpr _,
+  exact submodule.is_compl_orthogonal_of_complete_space.symm
```

Comment:
```suggestion
  refine K.orthogonal_family_self.is_hilbert_sum_internal _ _,
  refine le_trans _ (submodule.submodule_topological_closure _),
  rw supr_bool_eq,
  exact submodule.is_compl_orthogonal_of_complete_space.2
```
I propose getting rid of the lemma before and replacing this proof by the above.